File: codeeval/prefix-expressions.rb
# https://www.codeeval.com/open_challenges/7/
# You are given a prefix expression. Write a program which evaluates it. 

class String
  def polish
    if self.match( /[+*\/] \d+ \d+/ )
      astern  = $`.strip
      ahead   = $'.strip
      ops = $&.split(/ /)
      
      "#{astern} #{ops[1].to_i.send(ops[0].intern, ops[2].to_i)} #{ahead}".polish
    else
      self.strip
    end
  end
end

File.open(ARGV[0]).each_line do |line|
  puts line.chomp.polish
end

File: codeeval/prefix-expressions.py
import sys
import operator as op

with open(sys.argv[1], 'r') as f:
    lines = f.read().strip().splitlines()

for l in lines:
    ops = { "+": op.add, "*": op.mul, "/": op.truediv }

    s = []
    for i in l.split(' ')[::-1]:
        if i not in ops:
            s.append(int(i))
        else:
            result = ops[i](s.pop(), s.pop())
            s.append(result)
    print int(round(s.pop()))
